src/OpenCV/Features2d.hsc

```diff
@@ -625,50 +687,134 @@ newBFMatcher normType crossCheck = fromPtr
+                     Nothing
+    exceptErrorIO $ pureExcept $
+      withMatM (Proxy :: Proxy [height, width2])
+               (Proxy :: Proxy channels)
+               (Proxy :: Proxy depth)
+               white $ \imgM -> do
+        matCopyToM imgM (V2 0     0) frog        Nothing
+        matCopyToM imgM (V2 width 0) rotatedFrog Nothing
+
+        -- Draw the matches as lines from the query image to the train image.
+        forM_ matches $ \dmatch -> do
+          let matchRec = dmatchAsRec dmatch
+              queryPt = kpts1 V.! fromIntegral (dmatchQueryIdx matchRec)
+              trainPt = kpts2 V.! fromIntegral (dmatchTrainIdx matchRec)
+              queryPtRec = keyPointAsRec queryPt
+              trainPtRec = keyPointAsRec trainPt
+
+          -- We translate the train point one width to the right in order to
+          -- match the position of rotatedFrog in imgM.
+          line imgM
+               (round \<$> kptPoint queryPtRec :: V2 Int32)
+               ((round \<$> kptPoint trainPtRec :: V2 Int32) ^+^ V2 width 0)
+               blue 1 LineType_AA 0
+  where
+    orb = mkOrb defaultOrbParams {orb_nfeatures = 50}
+
+    width = fromInteger $ natVal (Proxy :: Proxy width)
+
+    rotatedFrog = exceptError $
+                  warpAffine frog rotMat InterArea False False (BorderConstant black)
+    rotMat = getRotationMatrix2D (V2 250 195 :: V2 CFloat) 45 0.8
+@
 
-        (numMatches :: Int) <- fromIntegral <$> peek numMatchesPtr
-        arrayPtr <- peek arrayPtrPtr
-        matches <- mapM (fromPtr . pure) =<< peekArray numMatches arrayPtr
+<<doc/generated/examples/fbMatcherImg.png fbMatcherImg>>
+
+<http://docs.opencv.org/3.0-last-rst/modules/features2d/doc/common_interfaces_of_descriptor_matchers.html#flannbasedmatcher OpenCV Sphinx doc>
+-}
+newtype FlannBasedMatcher = FlannBasedMatcher {unFlannBasedMatcher :: ForeignPtr C'FlannBasedMatcher}
+
+type instance C FlannBasedMatcher = C'FlannBasedMatcher
+
+instance WithPtr FlannBasedMatcher where
+    withPtr = withForeignPtr . unFlannBasedMatcher
+
+instance FromPtr FlannBasedMatcher where
+    fromPtr = objFromPtr FlannBasedMatcher $ \ptr ->
+                [CU.exp| void { delete $(FlannBasedMatcher * ptr) }|]
+
+
+--------------------------------------------------------------------------------
+
+
+data FlannIndexParams = FlannKDTreeIndexParams { trees :: Int }
+                      | FlannLshIndexParams { tableNumber :: Int, keySize :: Int, multiProbeLevel :: Int }
 
-        [CU.block| void {
-          delete [] *$(DMatch * * * arrayPtrPtr);
-        }|]
 
-        pure $ V.fromList matches
+data FlannSearchParams = FlannSearchParams { checks :: Int, eps :: Float, sorted :: Bool }
+
+
+-- NB: 1) it's OK to pass these new object as raw pointers because these directly pass to Ptr() in FlannBasedMatcher
+--     2) also, these objects use only in this internal module, so we don't create inlinec-wrappers for it, but pass
+--        between calls as void* pointers
+
```

Comment:
Please, look at  [FlannBasedMatcher constuctor](http://docs.opencv.org/trunk/dc/de2/classcv_1_1FlannBasedMatcher.html#ab9114a6471e364ad221f89068ca21382). It accepts two parameters, all are wrapped into `Ptr`.

So, when these params are created in `newFlannBasedMatcher` via `marshall*Params`:

```
newFlannBasedMatcher :: FlannBasedMatcherParams -> IO FlannBasedMatcher
newFlannBasedMatcher FlannBasedMatcherParams{..} = fromPtr
    [CU.exp|FlannBasedMatcher * {
      new cv::FlannBasedMatcher((flann::IndexParams*)($(void* c'indexParams)), (flann::SearchParams*)($(void* c'searchParams)))
    }|]
  where
    c'indexParams  = marshalIndexParams indexParams
    c'searchParams = marshallSearchParams searchParams

```

They are directly pass into proper `Ptr` constructor. So memory leak may occur only when first object was created, but creating of second object cast exception. Due to simplcity of second object this situation may occur only when memory leak. I think it very unlikely. So I did not write wrappers for these objects.

Am I convinced you? =)